core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -289,55 +296,58 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    * A linked list of watched delayed operations based on some key
    */
   private class Watchers(val key: Any) {
+    private[this] val operations = new ConcurrentLinkedQueue[T]()
+
+    // tracked separately because size() is O(n) for ConcurrentLinkedQueue
+    private[this] val numWatched = new AtomicInteger(0)
 
-    private[this] val operations = new LinkedList[T]()
+    def watched: Int = numWatched.get
 
-    def watched: Int = operations synchronized operations.size
+    def isEmpty: Boolean = operations.isEmpty
 
     // add the element to watch
     def watch(t: T) {
-      operations synchronized operations.add(t)
+      operations.add(t)
+      numWatched.incrementAndGet()
     }
 
     // traverse the list and try to complete some watched elements
     def tryCompleteWatched(): Int = {
-
-      var completed = 0
-      operations synchronized {
-        val iter = operations.iterator()
-        while (iter.hasNext) {
-          val curr = iter.next()
-          if (curr.isCompleted) {
-            // another thread has completed this operation, just remove it
-            iter.remove()
-          } else if (curr synchronized curr.tryComplete()) {
-            completed += 1
-            iter.remove()
-          }
-        }
+      var completedNow = 0
+      var completedAlready = 0
+
+      val iter = operations.iterator()
+      while (iter.hasNext) {
+        val curr = iter.next()
+        if (curr.isCompleted)
+          completedAlready += 1
```

Comment:
It probably does not matter, but I am wondering if it is better to remove the item here so we can potentially avoid an additional iteration? Are we doing this in the purge completed due to the concern of concurrent modification?